app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt

```diff
@@ -561,46 +579,33 @@ class StateFragmentPresenter @Inject constructor(
     pendingItemList: MutableList<StateItemViewModel>,
     hasPreviousState: Boolean,
     canContinueToNextState: Boolean,
     hasGeneralContinueButton: Boolean,
-    stateIsTerminal: Boolean
-  ) {
-    val stateNavigationButtonViewModel = StateNavigationButtonViewModel(context, this as StateNavigationButtonListener)
-    stateNavigationButtonViewModel.updatePreviousButton(isEnabled = hasPreviousState)
-
-    // Set continuation button.
+    stateIsTerminal: Boolean) {
     when {
-      hasGeneralContinueButton -> {
-        stateNavigationButtonViewModel.updateContinuationButton(
-          ContinuationNavigationButtonType.CONTINUE_BUTTON, isEnabled = true
-        )
-      }
-      canContinueToNextState -> {
-        stateNavigationButtonViewModel.updateContinuationButton(
-          ContinuationNavigationButtonType.NEXT_BUTTON, isEnabled = canContinueToNextState
-        )
-      }
-      stateIsTerminal -> {
-        stateNavigationButtonViewModel.updateContinuationButton(
-          ContinuationNavigationButtonType.RETURN_TO_TOPIC_BUTTON, isEnabled = true
-        )
-      }
-      viewModel.doesMostRecentInteractionRequireExplicitSubmission(pendingItemList) -> {
-        stateNavigationButtonViewModel.updateContinuationButton(
-          ContinuationNavigationButtonType.SUBMIT_BUTTON, isEnabled = true
-        )
-      }
-      else -> {
-        // No continuation button needs to be set since the interaction itself will push for answer submission.
-        stateNavigationButtonViewModel.updateContinuationButton(
-          ContinuationNavigationButtonType.CONTINUE_BUTTON, isEnabled = false
-        )
-      }
-    }
-    pendingItemList += stateNavigationButtonViewModel
+      hasGeneralContinueButton -> ContinueNavigationButtonViewModel(
+        hasPreviousState, fragment as PreviousNavigationButtonListener, fragment as ContinueNavigationButtonListener
+      )
+      canContinueToNextState -> NextButtonViewModel(
+        hasPreviousState, fragment as PreviousNavigationButtonListener, fragment as NextNavigationButtonListener
+      )
+      stateIsTerminal -> ReturnToTopicButtonViewModel(
+        hasPreviousState, fragment as PreviousNavigationButtonListener,
+        fragment as ReturnToTopicNavigationButtonListener
+      )
+      viewModel.doesMostRecentInteractionRequireExplicitSubmission(pendingItemList) -> SubmitButtonViewModel(
+        hasPreviousState, fragment as PreviousNavigationButtonListener, fragment as SubmitNavigationButtonListener
+      )
+      // Otherwise, just show the previous button since the interaction itself will push the answer submission.
+      hasPreviousState && !viewModel.isMostRecentInteractionAutoNavigating(pendingItemList) -> PreviousButtonViewModel(
+        fragment as PreviousNavigationButtonListener
+      )
+      // Otherwise, there's no navigation button that should be shown since the current interaction handles this.
+      else -> null
+    }?.let(pendingItemList::add)
   }
```

Comment:
@BenHenning I am getting various type mismatch errors in this file, like:
`Required: kotlin.String` and `Found: java.lang.String` in userAnswer.htmlAnswer.

The app is running perfectly but in android studio I am an seeing these errors.

Reply:
Are you still seeing those issues? It might be a codegen/AS out-of-sync issue. I'm not noticing it locally, so if you run into it again I suggest invalidating Android Studio's caches, running Gradle clean, then resyncing + rebuilding to see if the errors go away.